Weather Application using Spring WebFlux

maheshpardeshi-clairvoyant opened the pull request:
-- Consumed OpenWeather API to get weather details of city
-- Finding and storing user specific Cities to monitor weather
-- Implemented Caching using Caffeine cache for weather data of city

Commit:
Weather Application using Spring WebFlux
-- Consumed OpenWeather API to get weather details of city
-- Finding and storing user specific Cities to monitor weather
-- Implemented Caching using Caffeine cache for weather data of city

Commit:
Weather Application using Spring WebFlux
-- Fixed some review comments
- Removed unused router

ahmadimtcs commented on src/main/java/com/mahesh/weather/client/OpenWeatherRestClient.java:
```diff
@@ -0,0 +1,77 @@
+  @Value(value = "${key.openweather-url}")
+  private String OPENWEATHER_URL;
+
+  @Value(value = "${key.api-key}")
```
Use constructor injection

ahmadimtcs commented on src/main/java/com/mahesh/weather/client/OpenWeatherRestClient.java:
```diff
@@ -0,0 +1,77 @@
+public class OpenWeatherRestClient {
+
+  @Autowired
+  WebClient webClient;
```
this should be private

ahmadimtcs commented on src/main/java/com/mahesh/weather/exceptionhandler/BadDataException.java:
```diff
@@ -0,0 +1,12 @@
+
+public class BadDataException extends RuntimeException {
+
+  String message;
```
this should be private

ahmadimtcs commented on src/main/java/com/mahesh/weather/exceptionhandler/GlobalErrorHandler.java:
```diff
@@ -0,0 +1,31 @@
+import reactor.core.publisher.Mono;
+
+@Slf4j
+@Component
```
Please mark this as @ControllerAdvice to be more specific.

ahmadimtcs commented on src/main/java/com/mahesh/weather/handler/UserCityHandler.java:
```diff
@@ -0,0 +1,170 @@
+@Slf4j
+public class UserCityHandler {
+
+  @Autowired
```
Use constructor injection

ahmadimtcs commented on src/main/java/com/mahesh/weather/handler/UserCityHandler.java:
```diff
@@ -0,0 +1,170 @@
+          return cityService.findByNameAndCountry(StringUtils.capitalize(city.getName()),
+                  StringUtils.upperCase(city.getCountry()))
+              .switchIfEmpty(cityService.checkCityInOpenWeather(city));
+        }).collect(Collectors.toSet())
```
As discussed, we can make this flux and work on that.

ahmadimtcs commented on src/main/java/com/mahesh/weather/handler/UserCityHandler.java:
```diff
@@ -0,0 +1,170 @@
+              return userCityDetailsService.save(userCityDetails);
+            }))
+        .flatMap(userCityDetails -> ServerResponse.ok().bodyValue(userCityDetails))
+        .switchIfEmpty(Mono.error(new ResourceNotFoundException("City details are invalid")))
```
We can defer this call, otherwise it will be called every time

ahmadimtcs commented on src/main/java/com/mahesh/weather/handler/UserCityHandler.java:
```diff
@@ -0,0 +1,170 @@
+          log.info("City Name :=> {} Country:=>  {}", city.getName(), city.getCountry());
+          return cityService.findByNameAndCountry(StringUtils.capitalize(city.getName()),
+                  StringUtils.upperCase(city.getCountry()))
+              .switchIfEmpty(cityService.checkCityInOpenWeather(city));
```
We can defer this call, otherwise it will be called every time

ahmadimtcs reviewed (commented):
General comments:

- Always use constructor injection
- All the fields should be private unless it's absolutely necessary otherwise
- switchIfEmpty gets called every time, so we need to defer it.

Commit:
Implemented Review Comments added by Imteyaz
-- Used Mono.defer in switchIfEmpty
-- Switched to Constructor Injection pattern
-- Made fields in class as private
